libs/libcommon/src/libcommon/viewer_utils/asset.py

```diff
@@ -68,71 +77,97 @@ class ImageSource(TypedDict):
     width: int
 
 
-def create_image_file(
-    dataset: str,
-    config: str,
-    split: str,
-    row_idx: int,
-    column: str,
-    filename: str,
-    image: Image.Image,
-    assets_base_url: str,
-    assets_directory: StrPath,
-    overwrite: bool = True,
-) -> ImageSource:
-    dir_path, url_dir_path = create_asset_dir(
-        dataset=dataset,
-        config=config,
-        split=split,
-        row_idx=row_idx,
-        column=column,
-        assets_directory=assets_directory,
-    )
-    makedirs(dir_path, ASSET_DIR_MODE, exist_ok=True)
-    file_path = dir_path / filename
-    if overwrite or not file_path.exists():
-        image.save(file_path)
-    return {
-        "src": f"{assets_base_url}/{url_dir_path}/{filename}",
-        "height": image.height,
-        "width": image.width,
-    }
-
-
 class AudioSource(TypedDict):
     src: str
     type: str
 
 
-def create_audio_file(
+SupportedSource = Union[ImageSource, AudioSource]
+
+
+def upload_asset_file(
+    url_dir_path: str,
+    filename: str,
+    file_path: Path,
+    overwrite: bool = True,
+    s3_client: Optional[S3Client] = None,
+    s3_bucket: Optional[str] = None,
+    s3_folder_name: Optional[str] = None,
+) -> None:
+    if s3_client is not None and s3_bucket is not None:
+        object_key = f"{s3_folder_name}/{url_dir_path}/{filename}"
+        if overwrite or not s3_client.exists_in_bucket(s3_bucket, object_key):
+            s3_client.upload_to_bucket(str(file_path), s3_bucket, object_key)
+            os.remove(file_path)
+
+
+def create_asset_file(
     dataset: str,
     config: str,
     split: str,
     row_idx: int,
     column: str,
-    audio_file_bytes: bytes,
-    audio_file_extension: str,
-    assets_base_url: str,
     filename: str,
-    assets_directory: StrPath,
-    overwrite: bool = True,
-) -> list[AudioSource]:
-    dir_path, url_dir_path = create_asset_dir(
-        dataset=dataset,
-        config=config,
-        split=split,
-        row_idx=row_idx,
-        column=column,
-        assets_directory=assets_directory,
+    storage_options: Union[DirectoryStorageOptions, S3StorageOptions],
+    fn: Callable[[Path, str, bool], SupportedSource],
+) -> SupportedSource:
+    # get url dir path
+    assets_base_url = storage_options.assets_base_url
```

Comment:
It seems that it's the same as the one that points to the disk, is it expected ? (app_config.cached_assets.base_url)

Reply:
Yes, it will be the same as app_config.cached_assets.base_url (https://datasets-server.huggingface.co/cached-assets).
In this case, the prod DNS is not yet pointing to CloudFront for that URL so probably, the first test dataset images/audios won't be available via the public domain, just using the CloudFront https://d1j3nh2g9qznw6.cloudfront.net/cached-assets/*.
Once we confirm that everything is working as expected infra team will be able to update the Datasets Server DNS record to point on Cloudfront right @rtrompier?
I think for the first tests, we can at least ensure that the images/audio are being stored in S3 and are accessible via CloudFront.